text/0002-extensible-provisioning-templates.md

```diff
@@ -0,0 +1,118 @@
+-   Feature Name: Extensible provisioning templates
+-   Start Date: 2016-06-28
+-   RFC PR: (leave this empty)
+-   Issue: (leave this empty)
+
+# Summary
+
+[summary]: #summary
+
+This is a suggestion for refactoring Foreman's template generation mechanism is such a way that will enable
+different plugins to add plugin-specific code to the rendered template.
+
+# Motivation
+
+[motivation]: #motivation
+
+## Original use case
+
+As part of the effort to clean up Foreman's core from puppet related code,
+I have stumbled upon `Host::Managed#handle_ca` method. This method is responsible
+for dealing with certificate autosigning mechanism. The method is called from different parts
+in the code: from `UnattendedController` and from two orchestration concerns: `Orchestration::SSHProvision`
+and `Orchestration::Compute`. The motivation for this method is to keep the autosigning window open
+for as shortest period of time as possible.
+My suggestion is to solve this design flaw by using extensible templates in the following process:
+
+1.  Foreman will generate a certificate during the regular orchestration process and store it as part of
```

Comment:
It's isn't secure, but technically neither is autosign. Consider today's method:

1. Host GETs /unattended/provision
2. Orchestration of handle_ca is triggered
3. Autosign entry is put into the autosign file

At this point, any attacker watching the network traffic knows the name of the host being built, and can request the cert from the puppet master - it probably has about 15 mins to do so, given a reasonably average Kickstart process

In this approach, we render the certificate in the finish script, which is also requested via /unattended/finish - again, a watching attacker could use this info.

So, we're no *less* secure this way, but (a) we're can be *more* secure where https is possible (recent versions of Anaconda *do* support https), and (b) we make the orchestration of "things that need to happen on other servers" more pluggable.